lib/iris/config.py

```diff
@@ -154,3 +168,267 @@ def get_dir_option(section, option, default=None):
+
+    @property
+    def _defaults_dict(self):
+        """
+        Define the default value and available options for each settable
+        `kwarg` of this `Option`.
+
+        Note: `'options'` can be set to `None` if it is not reasonable to
+        specify all possible options. For example, this may be reasonable if
+        the `'options'` were a range of numbers.
+
+        """
+        return {'_scheduler': {'default': None, 'options': None},
+                'scheduler': {'default': 'threaded',
+                              'options': ['threaded',
+                                          'multiprocessing',
+                                          'async',
+                                          'distributed']},
+                'num_workers': {'default': 1, 'options': None},
+                'dask_scheduler': {'default': None, 'options': None},
+                }
+
+    def set__scheduler(self, value):
+        return value
+
+    def set_scheduler(self, value):
+        default = self._defaults_dict['scheduler']['default']
+        if value is None:
+            value = default
+        elif re.match(r'^(\d{1,3}\.){3}\d{1,3}:\d{1,5}$', value):
+            if DISTRIBUTED_AVAILABLE:
+                value = 'distributed'
+            else:
+                # Distributed not available.
+                wmsg = 'Cannot import distributed. Defaulting to {}.'
+                warnings.warn(wmsg.format(default))
+                self.set_scheduler(default)
+        elif value not in self._defaults_dict['scheduler']['options']:
+            # Invalid value for `scheduler`.
+            wmsg = 'Invalid value for scheduler: {!r}. Defaulting to {}.'
+            warnings.warn(wmsg.format(value, default))
+            self.set_scheduler(default)
+        return value
+
+    def set_num_workers(self, value):
+        default = self._defaults_dict['num_workers']['default']
+        scheduler = self.get('scheduler')
+        if scheduler == 'async' and value != default:
+            wmsg = 'Cannot set `num_workers` for the serial scheduler {!r}.'
+            warnings.warn(wmsg.format(scheduler))
+            value = None
+        elif scheduler == 'distributed' and value != default:
+            wmsg = ('Attempting to set `num_workers` with the {!r} scheduler '
+                    'requested. Please instead specify number of workers when '
+                    'setting up the distributed scheduler. See '
+                    'https://distributed.readthedocs.io/en/latest/index.html '
+                    'for more details.')
+            warnings.warn(wmsg.format(scheduler))
+            value = None
+        else:
+            if value is None:
+                value = default
+            if value >= cpu_count():
+                # Limit maximum CPUs used to 1 fewer than all available CPUs.
+                wmsg = ('Requested more CPUs ({}) than total available ({}). '
+                        'Limiting number of used CPUs to {}.')
```

Comment:
@dkillick Is this going to be a bit confusing?  What if somebody requests that maximum number of CPUs available, and then is only allowed to use one less than the max available, but the warning message states that the total available is the same number that they requested.  

For example, let's say my CPU count is 8, and as a parallelisation beginner I want to see how fast my script is when I run it on all 8 CPUs, so I request 8.  This isn't allowed by set_num_workers, so I get a warning message which says:
'Requested more CPUs (8) than total available (8).  Limiting number of used CPUs to 7'

I would find this message confusing, and it would make me angry.  Maybe it just needs a little clarification that you can't use all the CPUs available because it's silly.